### ISSUE TITLE
Decide: EXIF NN output contract (logits vs percentages)

### ISSUE THREAD

**Issue by victor-wildlife** [wildlifeai/Seeed_Grove_Vision_AI_Module_V2#166]
**What:** PR #141 enabled `USE_PERCENTAGE`, so the EXIF UserComment carries percentages and the website's ingestion consumes label + confidence in that form. The review argues for restoring raw logits: simpler, no floating point on the device, and the server can derive percentages at any time.

**Where:** `image_task.c`, `cvapp.cpp`, `exif_builder.c`; consumer: ww-website ingestion

**Evidence / context:** Cross-repo contract — see https://github.com/wildlifeai/Seeed_Grove_Vision_AI_Module_V2/blob/dev/_Documentation/development%20reports/2026-07_pr141-review-cgp/review_responses.md §7.

**Suggested fix:** Decide the device↔cloud encoding with the website work in view. Related: keep EXIF tag 0x9209 spec-compliant (the proposed 0/1/2 flash-type value belongs in the MakerNote), and remove `NN_confidence_EXIF_not_written.md` if logits win.

**Comment by victor-wildlife** [comment]
**What the website actually consumes** (checked in `ww-website` @ `e51b573`):

- `backend/app/domain/exif.py` reads **UserComment** (0x9286), parsing the
  `"label: value; label: value;"` payload into `user_comment_fields`, and **MakerNote**
  (0x927C) for the capture-settings CSV. It does **not** read `TAG_NN_DATA` — the only
  0xF2xx tag it knows is `0xF200` (Deployment_ID) — and it ignores the private
  0xF300–0xF3FF confidence tags.
- `backend/app/domain/edge_reflection.py` turns those fields into `observations` rows
  (`ai_origin='edge'` — the Camera AI result shown beside SpeciesNet). It matches
  **label names** against the project model's `ai_models.label_map`, compares the value
  with a per-label **threshold on a 0–100 scale**, and stores
  `classification_probability = pct / 100.0`.

**Consequences for this decision:**

1. The raw logits are *already* in every photo — `image_task.c` writes the NN output
   vector to `TAG_NN_DATA` unconditionally, so the percentage code only ever *added*
   UserComment and the 0xF3xx tags. This was never "logits or percentages".
2. But simply dropping `USE_PERCENTAGE` would silently stop edge observations appearing:
   the logits are a bare score array with no labels, and the website matches on names.

**Revised options:**

| | Option | Website work |
|---|---|---|
| a | Keep as-is | none |
| b | Emit **raw logits in the same `"label: value"` UserComment format** | small — rescale `label_map` thresholds and the `pct/100` conversion |
| c | Drop UserComment scores; website reads `TAG_NN_DATA` | substantial — parse the array *and* resolve labels from model metadata |

(b) reaches the actual goal — no floating-point code in the firmware, since logits are
`int8` — while the website's parser keeps working unchanged. Recommend scheduling (b) as a
coordinated change rather than doing it at merge time; merge #141 as-is meanwhile.

**Comment by victor-wildlife** [comment]
## Input from Charles Palmer (18 Aug 2026)

Recording this here so the decision has the reasoning attached to it rather than living in email.

### Background reading he suggests

Charles has read up on logits vs probability and recommends the same to whoever takes this on. He
did not send URLs, so as he described them:

1. A long post that gives the mathematical operation to convert one to the other, using logs and
   exponents.
2. "Are Logits a Confidence Score? Logits vs Probability", a low-maths explanation, including a
   warning that the percentage value might not be trustworthy.
3. "What Are Logits in AI? A Plain-English Explanation", linked from the above.

### His position, and why

From `doc/CGP_Code_Review_July26.md` and confirmed by email, the percentage code was Tobyn's. Charles
commented it out deliberately because:

- it is quite complex
- it is the only floating-point arithmetic in our codebase, and pulls in new libraries
- percentages are derived from logits, so the server can compute them at any time if wanted
- logits are the native TFLM output: simpler and more compact

His intention was to delete the percentage code entirely at some later point. He adds that he has
never timed `outputAsPercentage()` or measured its code size, and notes we may eventually run out of
memory. His summary: "This complexity, the use of floating point maths libraries and the concern that
the interpretation of percentages might be deceptive led to my judgement to omit percentages."

His conditional, from the review doc: **"Unless the app and website require percentages, I would like
to remove the percentage code."**

And: **"If we are to retain percentages there is merit in moving the code to its own C file."** That
is exactly #185, so this is Charles independently confirming that issue's premise.

### Reference documentation

`_Documentation/WW500_Neural_Network_Operation.md` (CGP, 22 April 2026) is the current description of
NN operation including the EXIF tags. The prompt that produced it, `CLAUDE_NN_operation.md`, exists
only on Charles's machine because `.gitignore` keeps it off GitHub. Worth noting that our `.gitignore`
is silently discarding the input half of AI-produced documentation while keeping the output.

---

## Verified against the code on `dev` (8cbb6a2b)

### Current state

`USE_PERCENTAGE` is **defined** at `cvapp.h:36`, so percentages are live today. That was not Charles's
doing: he had it commented out, and it was re-enabled during the AI-assisted work, which also made
`ENABLE_EXIF_CONFIDENCE` true and switched the UserComment tag from logits to percentages. The
comment block now above the define cites `doc/NN_confidence_EXIF_not_written.md`, a document Charles
has asked to be removed on the grounds that its central claim ("images carry no NN scores for the
website to ingest") is wrong: `TAG_NN_DATA` (0xC000) carries the raw logits unconditionally and
always has.

So today every photo carries **both**: logits in 0xC000 and percentages in UserComment.

### The website does require percentages, and will silently misread logits

This settles Charles's conditional, and it is the part I would not decide without.
`ww-website/backend/app/domain/edge_reflection.py` reflects on-device scores into `observations`:

```python
pct = _parse_pct(raw_value)                      # "87%" / "87" / 87 -> 87
if pct < threshold:                              # DEFAULT_REFLECT_THRESHOLD_PCT = 50
    continue
...
"classification_probability": pct / 100.0,
```

`_parse_pct` accepts a bare number. If the firmware reverts UserComment to logits, a logit of 71
parses as `pct = 71`, clears the default threshold of 50, and is written as
`classification_probability = 0.71`. Nothing errors. Negative logits fall below threshold and are
dropped, so the failure is not even visible as missing rows in the obvious cases.

`classification_probability` then flows into Camtrap DP export (`camtrapdp.py:649`), so wrong values
would reach published data.

**Whatever we decide, the firmware and website changes have to land together.** Reverting the
firmware alone is a silent data-corruption path, not a clean rollback.

### Percentages are not comparable between models

This supports Charles's "might be deceptive" concern with something concrete.
`outputAsPercentage()` dequantizes with the model's own output scale before softmax:

```c
logits_float[i] = (results[i] - zero_point) * scale;   // scale from output->params.scale
```

So the same int8 logits give different percentages depending purely on how the model was quantized.
Taking the logits from our own documented example (71 and -71):

| Output tensor scale | Resulting percentages |
|---|---|
| 0.00390625 (1/256, the value the code comment cites) | 63.5% / 36.5% |
| 0.0274 | 98.0% / 2.0% |

(My arithmetic, not measured on device. Worth confirming against a real model's
`output->params.scale`, which is also the cheapest way to check whether the 98%/2% in
`WW500_Neural_Network_Operation.md` is representative or was taken from a model with a very
different scale.)

A reviewer comparing "98%" from one model against "64%" from another would reasonably assume the
first is far more confident, when the underlying evidence can be identical. Logits have the same
problem in the other direction, but they at least look like raw numbers rather than inviting a
probability reading.

### Two defects in `outputAsPercentage()` regardless of the decision

1. **The decimal digit is wrong whenever rounding goes up.** `confidence_int` is the rounded value,
   so `confidence - confidence_int` is negative in that case, and the clamp turns it into `.0`:

   ```c
   int confidence_int  = (int)(confidence + 0.5f);                              // 98.7 -> 99
   int confidence_frac = (int)((confidence - confidence_int) * 10.0f + 0.5f);   // -2.5 -> -2
   if (confidence_frac < 0) confidence_frac = 0;                                // prints "99.0%"
   ```

   Console output only, since EXIF uses `confidence_int`.

2. **Two stack VLAs sized from the model.** `float logits_float[num_classes]` and
   `probabilities[num_classes]` come from `output->dims`, unclamped, while `MAX_CLASSES` is 16 and is
   only applied afterwards when storing. A model with a large output vector sizes these off the
   tensor, not off our limit.

---

## Still open, and broader than this issue

Charles raises four questions that the EXIF contract alone does not answer:

1. For each NN calculation, what should we send from the AI processor to the BLE processor?
   (Today: just `NN+` / `NN-`.)
2. What should either processor do with it? There is some storage in operational parameters, but
   that only works for models with 2 outputs.
3. What should go via LoRaWAN to the server?
4. What should go to the app? Live NN processing could be a useful demo.

These span the whole result pipeline rather than the EXIF tag, so I would rather not bury them in
this issue. Suggest a separate one, "Define the NN result pipeline: AI processor to BLE, LoRaWAN and
app", linked from here.

## Related

#185 (extract EXIF assembly and `USE_PERCENTAGE` out of `image_task.c`, which Charles endorses),
#187 (LightSource tag), #153 (EXIF Model tag on RP3).

**Comment by victor-wildlife** [comment]
## The Camtrap DP rationale, checked against the spec

Added context from Victor: the reason for converting logits to percentages was to follow Camtrap DP
so that photos recorded on the SD card can be read automatically by any camera-trap software. Worth
checking that rationale carefully, because it turns out to point somewhere slightly different.

### What the standard actually says

From [the spec](https://camtrap-dp.tdwg.org/data/#observations.classificationProbability),
`observations.classificationProbability` is:

> "Degree of certainty of the (most recent) classification. Expressed as a probability, with `1`
> being maximum certainty."

Type `number`, minimum `0`, maximum `1`. So Victor's instinct is right: the standard wants 0 to 1,
not 0 to 100.

**We already meet that.** `edge_reflection.py` writes `pct / 100.0`, and `camtrapdp.py:649` exports
`classification_probability` from there. Nothing is out of range today.

### Camtrap DP does not define an EXIF convention

This is the part that changes the argument. Camtrap DP is a dataset-level format: `deployments`,
`media` and `observations` as Frictionless Tabular Data Resources, plus a `datapackage.json`. Its
only reference to EXIF is the `media.exifData` field, which holds "EXIF data of the media file,
formatted as a valid JSON object", so EXIF is carried as extracted JSON *inside the package*.

There is no part of the standard that reads a JPEG's own EXIF. No third-party camera-trap tool will
read our `UserComment` and populate `classificationProbability` from it. Producing that field always
requires something to build the observations table, and for us that is `edge_reflection.py`.

So writing percentages into EXIF does not by itself make an SD card readable by camtrap software.
If that is the goal we actually want, the thing to build is a Camtrap DP package (the CSVs plus
`datapackage.json`), either generated on the device or produced from a card on import. That is a
different piece of work and probably worth its own issue.

**The consequence for this decision: Camtrap DP does not force the choice.** Logits or percentages
both work, as long as something converts to 0 to 1 somewhere, and that converter already exists.

### But server-side conversion is not currently possible

Charles's argument for logits is that "percentages are derived from logits, so the server can compute
them at any time if they were needed". That is true mathematically, and it is not true of what we
ship today.

Softmax has to run on *dequantized* logits, and dequantization needs the model's output quantization
parameters:

```c
logits_float[i] = (results[i] - zero_point) * scale;   // cvapp.cpp:852
```

Both come from `output->params` and **never leave `cvapp.cpp`**. They are not in `TAG_NN_DATA`
(which carries only length, class count and the int8 values), not in `UserComment`, not sent to the
BLE processor, not stored anywhere. And as the table in my previous comment shows, the scale is
exactly what decides whether logits of 71 and -71 mean 64% or 98%. A server given only the int8
values would have to guess the one number that determines the answer.

So "keep logits, convert server-side" is a viable option, but only with one of:

- **(a)** ship `scale` and `zero_point` alongside the logits in EXIF, or
- **(b)** record them per model in the website's model registry when a model is published.

(b) looks better to me. They are a fixed property of a model, so putting them in every photo is
waste, and the website already keeps per-model records keyed by `firmware_model_id` and
`version_number` with a `label_map`. Adding the quantization parameters there is a small change to
somewhere that already exists.

### What that adds up to

Option (b) appears to satisfy everything at once, which is worth stating plainly so the decision can
be made against it rather than in the abstract:

- Charles gets what he wants: no floating point on the device, no new maths libraries, the percentage
  code deleted, compact native TFLM output in EXIF.
- Camtrap DP gets a genuine 0 to 1 probability, computed exactly rather than approximated.
- The percentages become reproducible and auditable after the fact, instead of being baked into a
  photo by whatever the device computed at capture time.

Costs, stated honestly: website work to store the quantization parameters and do the softmax, and
the firmware and website changes must ship together, because of the silent misread described in my
previous comment. It also means a photo taken off an SD card no longer carries a human-readable
confidence, which matters if anyone inspects cards directly.

Still a decision for Victor and Charles, not something to settle here. But the Camtrap DP argument
does not require percentages in EXIF, which was the main thing holding the two positions apart.

**Comment by victor-wildlife** [comment]
Superseded by #189, which consolidates this decision with Charles's August input, the Camtrap DP check, the website coupling, the quantization-parameter finding and the wider pipeline questions. Closing here so there is one place to decide.
